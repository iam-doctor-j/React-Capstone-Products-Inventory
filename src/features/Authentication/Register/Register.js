import React from 'react';
import { Form, Field, withFormik } from 'formik';
import * as Yup from 'yup';
import Card from 'react-bootstrap/Card';
import { fromString } from 'uuidv4';
import { connect } from 'react-redux';
import { register, refresh } from '../AuthenticationSlice';
import routes from '../../../constants/routes';
import { toast } from 'react-toastify';

export const RegisterForm = (props) => {
    const {touched, errors, isSubmitting, error} = props;
    return(
        <div className="container">
        <div className="row justify-content-center">
            <Card className="col-12 col-md-8 col-lg-6">
                <Card.Body>
                    <Card.Title>
                        <div className="d-flex">
                            <span className="heading-text">Add new product</span>
                            <a className="ml-auto" onClick={() => {props.dispatch(refresh()); props.history.goBack()}} style={{cursor: 'pointer', color: 'var(--secondary)'}}><i className="fas fa-arrow-circle-left"></i></a>
                        </div>
                    </Card.Title>
                    <Form>
                        <div className="form-group">
                            <label className="label-text">Email</label>
                            <Field className="form-control" type="text" name="email" placeholder="Enter email" />
                            {touched.email && errors.email && <span className="label-text" style={{color: 'red'}}>{errors.email}</span>}
                        </div>
                        <div className="form-group">
                            <label className="label-text">Password</label>
                            <Field className="form-control" type="password" name="password" placeholder="Enter password" />
                            {touched.password && errors.password && <span className="label-text" style={{color: 'red'}}>{errors.password}</span>}
                        </div>
                        <div className="form-group">
                            <label className="label-text">First Name</label>
                            <Field className="form-control" type="text" name="firstName" placeholder="Enter First Name" />
                            {touched.firstName && errors.firstName && <span className="label-text" style={{color: 'red'}}>{errors.firstName}</span>}
                        </div>
                        <div className="form-group">
                            <label className="label-text">Last Name</label>
                            <Field className="form-control" type="text" name="lastName" placeholder="Enter Last Name" />
                            {touched.lastName && errors.lastName && <span className="label-text" style={{color: 'red'}}>{errors.lastName}</span>}
                        </div>
                        <div className="form-group">
                            <label className="label-text">Location</label>
                            <Field className="form-control" type="text" name="location" placeholder="Enter Location" />
                            {touched.location && errors.location && <span className="label-text" style={{color: 'red'}}>{errors.location}</span>}
                        </div>
                        <div className="form-group">
                            <label className="label-text">Mobile number</label>
                            <Field className="form-control" type="text" name="mobNumber" placeholder="Enter Mobile Number" />
                            {touched.mobNumber && errors.mobNumber && <span className="label-text" style={{color: 'red'}}>{errors.mobNumber}</span>}
                        </div>
                        <button className="btn btn-secondary" type="submit" disabled={isSubmitting}>{!isSubmitting ? <span>Register</span> : <span>Registering...</span>}</button>
                        <br/>
                        {error && <span className="label-text" style={{color: 'red'}}>{error}</span>}
                    </Form>
                </Card.Body>
            </Card>
        </div>
        </div>
    )};

const phoneRegExp = /^((\\+[1-9]{1,4}[ \\-]*)|(\\([0-9]{2,3}\\)[ \\-]*)|([0-9]{2,4})[ \\-]*)*?[0-9]{3,4}?[ \\-]*[0-9]{3,4}?$/

export const Register = withFormik({
    mapPropsToValues() {
        return {
            email: '',
            password: '',
            firstName: '',
            lastName: '',
            location: '',
            mobNumber: ''
        };
    },
    validationSchema: Yup.object().shape({
        email: Yup.string().email('Enter a valid email').required('This field is required'),
        password: Yup.string().required('This field is required'),
        firstName: Yup.string().min(3, 'First Name should be minimum 3 characters long').required('This field is required'),
        lastName: Yup.string().required('This field is required'),
        location: Yup.string().min(3, 'Location should be atleast 3 characters long').required('This field is required'),
        mobNumber: Yup.string().matches(phoneRegExp, 'Enter a valid phone number').required('This field is required')
    }),
    handleSubmit(values, { setSubmitting, props }) {

        let user = {
            id: fromString(JSON.stringify(values)+Date.now().toString()),
            email: values.email,
            password: values.password,
            firstName: values.firstName,
            lastName: values.lastName,
            location: values.location,
            mobNumber: values.mobNumber,
        }
        props.dispatch(register(user, () => {
            setSubmitting(false);
            toast.success('Registration successfull!');
            props.history.replace(routes.HOME);
        }, () => {
            setSubmitting(false);
            toast.error('Registration failed');
        }));
    }
})(RegisterForm);

export default connect(state => {
    
    return {
        error: state.auth.error,
    }
})(Register);