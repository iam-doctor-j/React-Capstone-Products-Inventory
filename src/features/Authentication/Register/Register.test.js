import React from 'react';
import { mount } from 'enzyme';
import { Register, RegisterForm } from './Register';
import Card from 'react-bootstrap/Card';

describe('Testing Register using Full DOM rendering', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = mount(<Register/>);
    })

    it('should render properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render RegisterForm', () => {
        expect(wrapper.find(RegisterForm).length).toEqual(1);
    })

    it('should render one Card inside RegisterForm', () => {
        expect(wrapper.find(RegisterForm).find(Card).length).toEqual(1);
    })

    afterEach(() => {
        wrapper.unmount();
    })
});