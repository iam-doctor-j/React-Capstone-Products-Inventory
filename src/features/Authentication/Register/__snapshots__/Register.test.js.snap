// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Register using Full DOM rendering should render properly 1`] = `
<WithFormik(RegisterForm)>
  <Formik
    handleSubmit={[Function]}
    initialValues={
      Object {
        "email": "",
        "firstName": "",
        "lastName": "",
        "location": "",
        "mobNumber": "",
        "password": "",
      }
    }
    onSubmit={[Function]}
    validationSchema={[Function]}
  >
    <RegisterForm
      dirty={false}
      errors={Object {}}
      getFieldHelpers={[Function]}
      getFieldMeta={[Function]}
      getFieldProps={[Function]}
      handleBlur={[Function]}
      handleChange={[Function]}
      handleReset={[Function]}
      handleSubmit={[Function]}
      initialErrors={Object {}}
      initialTouched={Object {}}
      initialValues={
        Object {
          "email": "",
          "firstName": "",
          "lastName": "",
          "location": "",
          "mobNumber": "",
          "password": "",
        }
      }
      isSubmitting={false}
      isValid={true}
      isValidating={false}
      registerField={[Function]}
      resetForm={[Function]}
      setErrors={[Function]}
      setFieldError={[Function]}
      setFieldTouched={[Function]}
      setFieldValue={[Function]}
      setFormikState={[Function]}
      setStatus={[Function]}
      setSubmitting={[Function]}
      setTouched={[Function]}
      setValues={[Function]}
      submitCount={0}
      submitForm={[Function]}
      touched={Object {}}
      unregisterField={[Function]}
      validateField={[Function]}
      validateForm={[Function]}
      validateOnBlur={true}
      validateOnChange={true}
      validateOnMount={false}
      values={
        Object {
          "email": "",
          "firstName": "",
          "lastName": "",
          "location": "",
          "mobNumber": "",
          "password": "",
        }
      }
    >
      <div
        className="container"
      >
        <div
          className="row justify-content-center"
        >
          <Card
            body={false}
            className="col-12 col-md-8 col-lg-6"
          >
            <div
              className="col-12 col-md-8 col-lg-6 card"
            >
              <CardBody>
                <div
                  className="card-body"
                >
                  <CardTitle>
                    <ForwardRef
                      className="card-title"
                    >
                      <div
                        className="card-title h5"
                      >
                        <div
                          className="d-flex"
                        >
                          <span
                            className="heading-text"
                          >
                            Add new product
                          </span>
                          <a
                            className="ml-auto"
                            onClick={[Function]}
                            style={
                              Object {
                                "color": "var(--secondary)",
                                "cursor": "pointer",
                              }
                            }
                          >
                            <i
                              className="fas fa-arrow-circle-left"
                            />
                          </a>
                        </div>
                      </div>
                    </ForwardRef>
                  </CardTitle>
                  <Form>
                    <form
                      action="#"
                      onReset={[Function]}
                      onSubmit={[Function]}
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          className="label-text"
                        >
                          Email
                        </label>
                        <Field
                          className="form-control"
                          name="email"
                          placeholder="Enter email"
                          type="text"
                        >
                          <input
                            className="form-control"
                            name="email"
                            onBlur={[Function]}
                            onChange={[Function]}
                            placeholder="Enter email"
                            type="text"
                            value=""
                          />
                        </Field>
                      </div>
                      <div
                        className="form-group"
                      >
                        <label
                          className="label-text"
                        >
                          Password
                        </label>
                        <Field
                          className="form-control"
                          name="password"
                          placeholder="Enter password"
                          type="password"
                        >
                          <input
                            className="form-control"
                            name="password"
                            onBlur={[Function]}
                            onChange={[Function]}
                            placeholder="Enter password"
                            type="password"
                            value=""
                          />
                        </Field>
                      </div>
                      <div
                        className="form-group"
                      >
                        <label
                          className="label-text"
                        >
                          First Name
                        </label>
                        <Field
                          className="form-control"
                          name="firstName"
                          placeholder="Enter First Name"
                          type="text"
                        >
                          <input
                            className="form-control"
                            name="firstName"
                            onBlur={[Function]}
                            onChange={[Function]}
                            placeholder="Enter First Name"
                            type="text"
                            value=""
                          />
                        </Field>
                      </div>
                      <div
                        className="form-group"
                      >
                        <label
                          className="label-text"
                        >
                          Last Name
                        </label>
                        <Field
                          className="form-control"
                          name="lastName"
                          placeholder="Enter Last Name"
                          type="text"
                        >
                          <input
                            className="form-control"
                            name="lastName"
                            onBlur={[Function]}
                            onChange={[Function]}
                            placeholder="Enter Last Name"
                            type="text"
                            value=""
                          />
                        </Field>
                      </div>
                      <div
                        className="form-group"
                      >
                        <label
                          className="label-text"
                        >
                          Location
                        </label>
                        <Field
                          className="form-control"
                          name="location"
                          placeholder="Enter Location"
                          type="text"
                        >
                          <input
                            className="form-control"
                            name="location"
                            onBlur={[Function]}
                            onChange={[Function]}
                            placeholder="Enter Location"
                            type="text"
                            value=""
                          />
                        </Field>
                      </div>
                      <div
                        className="form-group"
                      >
                        <label
                          className="label-text"
                        >
                          Mobile number
                        </label>
                        <Field
                          className="form-control"
                          name="mobNumber"
                          placeholder="Enter Mobile Number"
                          type="text"
                        >
                          <input
                            className="form-control"
                            name="mobNumber"
                            onBlur={[Function]}
                            onChange={[Function]}
                            placeholder="Enter Mobile Number"
                            type="text"
                            value=""
                          />
                        </Field>
                      </div>
                      <button
                        className="btn btn-secondary"
                        disabled={false}
                        type="submit"
                      >
                        <span>
                          Register
                        </span>
                      </button>
                      <br />
                    </form>
                  </Form>
                </div>
              </CardBody>
            </div>
          </Card>
        </div>
      </div>
    </RegisterForm>
  </Formik>
</WithFormik(RegisterForm)>
`;
